### ISSUE TITLE
False positive with NewType and singledispatch

### ISSUE THREAD

**Issue by guillaume-alliander** [microsoft/pyright#10447]
**Describe the bug**
Using `newType` as discriminator for `singledispatch` does not raise a error from pyright, but fails at runtime.
I would expect the same sort of error message as for when `isinstance` is used with a `NewType`.

**Code or Screenshots**

Code with both expected (for `isinstance`) and unexpected (for `singledispatch`) behaviour. 

```python
from functools import singledispatch
from typing import Any, NewType, overload

AnotherStr = NewType("AnotherStr", str)


@overload
def say_hello(arg: str) -> None: ...
@overload
def say_hello(arg: AnotherStr) -> None: ...


def say_hello(arg: str | AnotherStr) -> None:
    if isinstance(arg, AnotherStr):  # pyright righly complains about this
        print(f"Hello {arg}")
    else:
        print("Hello stranger")


@singledispatch
def another_hello(_: Any) -> None:
    print("Hello")


@another_hello.register
def _(arg: AnotherStr) -> None:  # Pyright does not complain, but there is a runtime error: TypeError: Invalid annotation for 'arg'. __main__.AnotherStr is not a class.
    print(f"Hello {arg}")

```

**VS Code extension or command-line**

On the cli, pyright 1.1.400, python 3.12.10

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

Pyright has no specialized knowledge of `functools.singledispatch`. It's analyzing the code using the type information provided in the `functools.pyi`. In general, pyright has no specialized knowledge of any functions even if they're in stdlib. The `isinstance` function is one of those rare exceptions. 

It would be possible to add custom logic that understands the details of singledispatch, but that's a high bar. [This issue](https://github.com/microsoft/pyright/issues/4472) is a feature request for supporting this feature, and so far it has received zero upvotes.
